File: src/main/java/com/parking/smartparking/service/SlotService.java
package com.parking.smartparking.service;

import com.parking.smartparking.entity.Slot;
import com.parking.smartparking.enums.SlotStatus;
import com.parking.smartparking.enums.VehicleType;
import com.parking.smartparking.exception.ResourceNotFoundException;
import com.parking.smartparking.repository.SlotRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;

/**
 * Service for managing parking slot operations.
 * Provides CRUD operations along with availability queries by location and vehicle type.
 */
@Service
@RequiredArgsConstructor
public class SlotService {

    private final SlotRepository slotRepository;

    /**
     * Retrieves all parking slots.
     *
     * @return list of all slots
     */
    public List<Slot> getAllSlots() {
        return slotRepository.findAll();
    }

    /**
     * Retrieves a slot by its unique identifier.
     *
     * @param id the slot ID
     * @return the Slot entity
     * @throws ResourceNotFoundException if no slot is found
     */
    public Slot getSlotById(Long id) {
        return slotRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Slot not found with id: " + id));
    }

    /**
     * Retrieves all slots belonging to a specific parking location.
     *
     * @param locationId the parking location ID
     * @return list of slots at the given location
     */
    public List<Slot> getSlotsByLocation(Long locationId) {
        return slotRepository.findByParkingLocationId(locationId);
    }

    /**
     * Retrieves all currently available slots across all locations.
     *
     * @return list of available slots
     */
    public List<Slot> getAvailableSlots() {
        return slotRepository.findByStatus(SlotStatus.AVAILABLE);
    }

    /**
     * Retrieves available slots at a specific parking location.
     *
     * @param locationId the parking location ID
     * @return list of available slots at the location
     */
    public List<Slot> getAvailableSlotsByLocation(Long locationId) {
        return slotRepository.findByParkingLocationIdAndStatus(locationId, SlotStatus.AVAILABLE);
    }

    /**
     * Retrieves available slots at a location filtered by vehicle type.
     *
     * @param locationId  the parking location ID
     * @param vehicleType the type of vehicle
     * @return list of matching available slots
     */
    public List<Slot> getAvailableSlotsByLocationAndType(Long locationId, VehicleType vehicleType) {
        return slotRepository.findByParkingLocationIdAndVehicleTypeAndStatus(
                locationId, vehicleType, SlotStatus.AVAILABLE);
    }

    /**
     * Creates a new parking slot.
     *
     * @param slot the slot to create
     * @return the saved Slot entity
     */
    public Slot createSlot(Slot slot) {
        slot.setCreatedAt(LocalDateTime.now());
        return slotRepository.save(slot);
    }

    /**
     * Updates an existing parking slot.
     *
     * @param id   the ID of the slot to update
     * @param slot the slot object with updated fields
     * @return the updated Slot entity
     * @throws ResourceNotFoundException if no slot is found
     */
    public Slot updateSlot(Long id, Slot slot) {
        Slot existingSlot = slotRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Slot not found with id: " + id));

        existingSlot.setSlotNumber(slot.getSlotNumber());
        existingSlot.setVehicleType(slot.getVehicleType());
        existingSlot.setStatus(slot.getStatus());

        return slotRepository.save(existingSlot);
    }

    /**
     * Deletes a slot by its unique identifier.
     *
     * @param id the ID of the slot to delete
     * @throws ResourceNotFoundException if no slot is found
     */
    public void deleteSlot(Long id) {
        Slot slot = slotRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Slot not found with id: " + id));
        slotRepository.delete(slot);
    }

    /**
     * Updates the status of a specific slot.
     *
     * @param id     the slot ID
     * @param status the new slot status
     * @return the updated Slot entity
     * @throws ResourceNotFoundException if no slot is found
     */
    public Slot updateSlotStatus(Long id, SlotStatus status) {
        Slot slot = slotRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Slot not found with id: " + id));
        slot.setStatus(status);
        return slotRepository.save(slot);
    }

    /**
     * Counts the number of slots with a given status.
     *
     * @param status the slot status to count
     * @return the count of slots with the specified status
     */
    public Long countByStatus(SlotStatus status) {
        return slotRepository.countByStatus(status);
    }
}

File: angular-frontend/src/app/services/slot.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ApiResponse } from '../models/api-response.model';

@Injectable({ providedIn: 'root' })
export class SlotService {
  private apiUrl = 'http://localhost:8080/api';

  constructor(private http: HttpClient) {}

  getAllSlots(): Observable<ApiResponse> {
    return this.http.get<ApiResponse>(`${this.apiUrl}/slots`);
  }

  getSlotById(id: number): Observable<ApiResponse> {
    return this.http.get<ApiResponse>(`${this.apiUrl}/slots/${id}`);
  }

  getSlotsByLocation(locationId: number): Observable<ApiResponse> {
    return this.http.get<ApiResponse>(`${this.apiUrl}/slots/location/${locationId}`);
  }

  getAvailableSlots(): Observable<ApiResponse> {
    return this.http.get<ApiResponse>(`${this.apiUrl}/slots/available`);
  }

  getAvailableSlotsByLocation(locationId: number): Observable<ApiResponse> {
    return this.http.get<ApiResponse>(`${this.apiUrl}/slots/available/${locationId}`);
  }

  getAvailableSlotsByLocationAndType(locationId: number, type: string): Observable<ApiResponse> {
    return this.http.get<ApiResponse>(`${this.apiUrl}/slots/available/${locationId}/${type}`);
  }

  createSlot(slot: any, locationId: number): Observable<ApiResponse> {
    return this.http.post<ApiResponse>(`${this.apiUrl}/slots?locationId=${locationId}`, slot);
  }

  updateSlot(id: number, slot: any): Observable<ApiResponse> {
    return this.http.put<ApiResponse>(`${this.apiUrl}/slots/${id}`, slot);
  }

  deleteSlot(id: number): Observable<ApiResponse> {
    return this.http.delete<ApiResponse>(`${this.apiUrl}/slots/${id}`);
  }

  // Location APIs
  getAllLocations(): Observable<ApiResponse> {
    return this.http.get<ApiResponse>(`${this.apiUrl}/locations`);
  }

  getActiveLocations(): Observable<ApiResponse> {
    return this.http.get<ApiResponse>(`${this.apiUrl}/locations/active`);
  }

  createLocation(location: any): Observable<ApiResponse> {
    return this.http.post<ApiResponse>(`${this.apiUrl}/locations`, location);
  }

  updateLocation(id: number, location: any): Observable<ApiResponse> {
    return this.http.put<ApiResponse>(`${this.apiUrl}/locations/${id}`, location);
  }

  deleteLocation(id: number): Observable<ApiResponse> {
    return this.http.delete<ApiResponse>(`${this.apiUrl}/locations/${id}`);
  }
}
